Add program analysis tool

tohrnii opened the pull request:
Partly addresses #138 

Adds a utility package to analyse any given script. Takes a source script as a file input and prints the following information.
- [X] Number of VM cycles it takes to execute the script
- [X] Number of NOOPs executed as part of a program

Input:
`proc.foo.1 pop.local.0 end begin popw.mem.1 push.17 exec.foo end`

Output:
```
Total Number of VM Cycles: 24
Total Number of NOOPs: 1
```

bobbinth commented:
@tohrnii - thank you for working on this! A few preliminary thoughts:

1. I'd like to make the structure a bit more generic. For example:
  - Let's return something like `ProgramInfo` struct from the `analyze()` function. For now, this could contain just a single field and we can expand it later to contain more fields.
  - I'd like the structure to be generic so that we could use it, for example, from WASM. This way, we'll be able to have a webpage where you could paste in a script and get the stats back.
2. I'd probably prefer `tools` rather than `analysis` crate. And the package name itself could be `miden-vm-tools`.
3. I'm not sure what's the right place to put the code is. On the one hand, it would be nice to have `miden-vm-tools` compile into a binary (so that we could run it from the command line). On the other hand, I'd like to be able to compile it as a WASM-compliant library. Any suggestions are welcome here.

grjte commented:
>     3. I'm not sure what's the right place to put the code is. On the one hand, it would be nice to have `miden-vm-tools` compile into a binary (so that we could run it from the command line). On the other hand, I'd like to be able to compile it as a WASM-compliant library. Any suggestions are welcome here.

I think it should probably be a library which could then be used for front-end tooling or as a dependency for a binary

bobbinth commented:
> I think it should probably be a library which could then be used for front-end tooling or as a dependency for a binary

I guess the question is where to put the library portion of the code and where to put the binaries. For example, we could put the actual code into the `miden` crate and then create `tools` crate to contain binaries.

Another option is for `tools` crate be a library crate and come up with some other crate to contain binaries.

tohrnii commented:
> I guess the question is where to put the library portion of the code and where to put the binaries. For example, we could put the actual code into the miden crate and then create tools crate to contain binaries.
This seems like the best way to go to me. Would it make sense to put the library in `core/utils`?

tohrnii commented:
Moved the lib to `miden::utils` and bin to `tools` crate. Does this structure make sense?

bobbinth reviewed (changes requested):
Looks good! I left a few comments inline.

bobbinth commented:
@tohrnii - yes, this is mostly likely due to #212. The difference in count should come primarily form more `NOOP` operations executed. We need to execute `NOOP`s in a few cases, for example:
1. If a number of operation groups in a batch is not a power of two, we execute `NOOP`s for each missing group (e.g., if there are 3 groups in a batch, we put a single `NOOP` into the 4th group).
2. If a `PUSH` operation is a last one in a group, we need to execute a `NOOP` operation after it.

By the way, it may be a good metric to track: number of `NOOP`s executed as a part of a program. This could be a good efficiency metric.

tohrnii commented:
Thanks for the confirmation on PR 212. I'll add NOOPs metric to be added as part of the program analysis tool to the original issue #138 as a comment.

bobbinth commented:
@tohrnii - let's add `NOOP` tracking in this PR (since I think it should be fairly simple), and then we can do the rest in subsequent PRs. What do you think?

tohrnii commented:
@bobbinth - Sounds good!

bobbinth commented:
Actually, I should say that counting NOOPs would be pretty difficult right now, but will become very easy once #195 is merges. So, let's wait until I'm done with decoder trace, and then we can finish this PR.

tohrnii commented:
@bobbinth - I've added `total noops executed` as part of the analysis. Also added None operation as part of this PR. VmState at the first row contains the None operation based on your suggestion.

Two pending analysis left to be addressed are:
- [ ] Number of program blocks in the script (for each block type)
- [ ] Number of cycles per assembly instruction. E.g., a program executed u32add.full operation x times, and this translated to y VM cycles.

I will address these in a separate PR after completing #226.

bobbinth commented on processor/src/decoder/mod.rs:
```diff
@@ -413,7 +413,11 @@ impl Decoder {
     /// Get operation at a particular clock cycle. Only applicable in debug mode.
     pub fn get_operation_at(&self, clk: usize) -> Operation {
-        self.operations[clk]
+        if clk == 0 {
+            Operation::None
+        } else {
+            self.operations[clk - 1]
+        }
```
I wouldn't put this logic into the decoder. Let's put it instead into the `VmStateIterator`.

bobbinth commented on core/src/operations/mod.rs:
```diff
@@ -381,6 +381,9 @@ pub enum Operation {
     Advice(AdviceInjector),
+
+    /// Added to the top of VmStateIterator
+    None,
```
I'm not too sure about having this here. It isn't really needed for VM execution - only for debug purposes. I was thinking that we'd change `VmState` to something like:
```Rust
pub struct VmState {
    pub clk: usize,
    pub op: Option<Operation>,
    pub fmp: Felt,
    pub stack: Vec<Felt>,
    pub memory: Vec<(u64, Word)>,
}
```
And handle `None` at that level.

tohrnii replied:
Thanks. I was unsure of what the right way to handle it was out of the 2 (introducing variant or changing type to Option). I've changed to Option based on your suggestion in the latest commit. Makes sense since it is not required in non-debug mode.

bobbinth commented on processor/src/operations/mod.rs:
```diff
@@ -137,6 +137,8 @@ impl Process {
             Operation::Debug(options) => self.op_debug(options)?,
             Operation::Advice(injector) => self.op_advice(injector)?,
+
+            Operation::None => (),
```
If we do end up keeping `None` as an operation variant, this should probably panic.

tohrnii replied:
Removed it in latest commit.

bobbinth commented on miden/src/tools/mod.rs:
```diff
@@ -0,0 +1,62 @@
+/// Returns program analysis for a given script string. Returns ProgramInfo with following fields:
+/// - total_vm_cycles (vm cycles it takes to execute the entire script)
+/// - total_noops (total noops executed as part of a program)
+pub fn analyze(program: String) -> ProgramInfo {
```
I think this should probably take `ProgramInputs` as a parameter for programs which cannot be executed against blank inputs.

tohrnii replied:
For the binary, do we need to allow `advice_tape` and `advice_sets` as input as well? Currently it only takes optional `stack_init` as external input from user.

bobbinth replied:
I mentioned this in a comment below, but let's not worry about this for now.

bobbinth reviewed (changes requested):
Thank you! I left a few more comments inline.

bobbinth commented on miden/src/tools/mod.rs:
```diff
@@ -0,0 +1,61 @@
+}
+
+impl ProgramInfo {
+    ///Creates a new ProgramInfo object
```
nit: missing space after `///`.

bobbinth commented on miden/src/tools/mod.rs:
```diff
@@ -0,0 +1,61 @@
+        }
+    }
+
+    ///Get total vm cycles to execute a program
```
nit: missing space after `///`.

bobbinth commented on miden/src/tools/mod.rs:
```diff
@@ -0,0 +1,61 @@
+        self.total_vm_cycles
+    }
+
+    ///Get total noops executed as part of a program
```
nit: missing space after `///`.

bobbinth commented on miden/src/tools/mod.rs:
```diff
@@ -0,0 +1,61 @@
+    let mut noop_count = 0;
+
+    for vm_state in vm_state_iterator {
+        if vm_state.as_ref().unwrap().op == Some(Operation::Noop) {
```
I think you should be able to change this to: `if matches!(op, Operation::Noop) {`.

tohrnii replied:
Thanks. Done.

bobbinth commented on tools/src/main.rs:
```diff
@@ -0,0 +1,36 @@
+    //reads input masm file
+    let program = std::fs::read_to_string(&args.masm_path).expect("Could not read masm file");
+    let mut program_inputs = ProgramInputs::none();
+    if args.stack_input_path.is_some() {
+        //reads input stack file
+        let stack_input_str = std::fs::read_to_string(&args.stack_input_path.unwrap())
+            .expect("Could not read stack input file");
+        let stack_input: Vec<u64> = stack_input_str
+            .split(' ')
+            .map(|x| x.parse::<u64>().unwrap())
+            .collect();
+        program_inputs = ProgramInputs::new(&stack_input, &[], vec![]).unwrap();
+    }
```
Let's not worry about stack inputs in this PR. There are a number of different ways to do this, and I think we should spend a bit more time on this.

bobbinth reviewed (changes requested):
Thank you! A few more comments inline.

tohrnii commented on processor/src/decoder/mod.rs:
```diff
@@ -413,7 +413,7 @@ impl Decoder {
     /// Get operation at a particular clock cycle. Only applicable in debug mode.
     pub fn get_operation_at(&self, clk: usize) -> Operation {
-        self.operations[clk]
+        self.operations[clk - 1]
```
Thanks. Done.

tohrnii commented:
@bobbinth - Thank you for the review. Made suggested changes.

bobbinth commented on miden/src/tools/mod.rs:
```diff
@@ -0,0 +1,61 @@
+    let mut noop_count = 0;
+
+    for vm_state in vm_state_iterator {
+        if matches!(vm_state.as_ref().unwrap().op, Some(Operation::Noop)) {
```
Do you need the `unwrap()` here? Would something like this work: `matches!(vm_state.op, Some(Operation::Noop))`?

bobbinth replied:
Oh - I think it won't. I'm trying to think of a way to get rid of the `unwrap`'s here - ideally, we shouldn't have them in production code.

bobbinth replied:
I guess one thing that this doesn't account for is when a program terminates with an error. We should probably change the return type of `analyze()` to return a `Result<ProgramInfo, ExecutionError>` and then if we encounter an error it would be just returned instead of `ProgramInfo`. What do you think?

tohrnii replied:
> I'm trying to think of a way to get rid of the unwrap's here - ideally, we shouldn't have them in production code.
Just to clarify, the reason is proper error propagation right?
Can we use something like:
```
for state in vm_state_iterator {
        let vm_state = match state {
            Ok(vm_state) => vm_state,
            Err(err) => return Err(err)
        };
        if matches!(vm_state.op, Some(Operation::Noop)) {
            noop_count += 1;
        }
        total_vm_cycles = vm_state.clk;
    }
```

tohrnii replied:
> I guess one thing that this doesn't account for is when a program terminates with an error. We should probably change the return type of analyze() to return a Result<ProgramInfo, ExecutionError> and then if we encounter an error it would be just returned instead of ProgramInfo. What do you think?

Makes sense to propagate error here. Since there are two possible errors here, `AssemblyError` while compiling the script and `ExecutionError` while executing the script, should we add a new enum called `AnalysisError` and return that instead? That doesn't look very clean though. Is there a better (more idiomatic) way to handle that? Thanks.

bobbinth replied:
Yes - the reason is proper error propagation, but also, as a general rule, I try not to have `unwrap`'s in non-test code.

Yes, let's add a new enum. We did something like this [here](https://github.com/maticnetwork/miden/blob/next/miden/tests/integration/helpers/mod.rs#L19). Though I don't love the name `AnalysisError`. One other option that came to mind is `MidenError` - but I don't love this one too much either.

The code snippet above looks right - though, I'd prefer to use [map_err()](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err) instead of `match`. For example, something like:
```Rust
for state in vm_state_iterator {
    let vm_state = state.map_err(...)?;
    if matches!(vm_state.op, Some(Operation::Noop)) {
        noop_count += 1;
    }
    total_vm_cycles = vm_state.clk;
}
```

bobbinth replied:
Maybe `ProgramError` is better than the other options?

tohrnii replied:
@bobbinth - `ProgramError` sounds good to me. In general, do you prefer more verbose names like if this enum is only going to be used for program analysis, then `ProgramAnalysisError` would be more verbose but not very clean or should we lean more towards succinctness.

tohrnii commented:
@bobbinth - Thank you for the review. Made suggested changes.
`ProgramError` enum variants take error as argument instead of string. The thinking is that this would allow access to properties of the error in case this enum was to be used outside of program analysis and in general to avoid loss of information. Please let me know if this should take a string instead.

bobbinth reviewed (approved):
All looks good! Thank you!
